default/AI/ResearchAI.py

```diff
@@ -71,481 +446,63 @@ def generate_research_orders():
-        unlocked_parts = []
-        for item in unlocked_items:
-            if item.type == fo.unlockableItemType.shipPart:
-                print "Tech %s unlocks a ShipPart: %s" % (tech, item.name)
-                unlocked_parts.append(item.name)
-            elif item.type == fo.unlockableItemType.shipHull:
-                print "Tech %s unlocks a ShipHull: %s" % (tech, item.name)
-                unlocked_hulls.append(item.name)
-        if not (unlocked_parts or unlocked_hulls):
-            print "No new ship parts/hulls unlocked by tech %s" % tech
-            continue
-        old_designs = ShipDesignAI.MilitaryShipDesigner().optimize_design(consider_fleet_count=False)
-        new_designs = ShipDesignAI.MilitaryShipDesigner().optimize_design(additional_hulls=unlocked_hulls, additional_parts=unlocked_parts, consider_fleet_count=False)
-        if not (old_designs and new_designs):
-            # AI is likely defeated; don't bother with logging error message
-            continue
-        old_rating, old_pid, old_design_id, old_cost = old_designs[0]
-        old_design = fo.getShipDesign(old_design_id)
-        new_rating, new_pid, new_design_id, new_cost = new_designs[0]
-        new_design = fo.getShipDesign(new_design_id)
-        if new_rating > old_rating:
-            print "Tech %s gives access to a better design!" % tech
-            print "old best design: Rating %.5f" % old_rating
-            print "old design specs: %s - " % old_design.hull, list(old_design.parts)
-            print "new best design: Rating %.5f" % new_rating
-            print "new design specs: %s - " % new_design.hull, list(new_design.parts)
-        else:
-            print "Tech %s gives access to new parts or hulls but there seems to be no military advantage." % tech
-
-
-
-
-def generate_default_research_order():
-    """
-    Generate default research orders.
-    Add cheapest technology from possible researches
-    until current turn point totally spent.
-    """
-
-    empire = fo.getEmpire()
+    research_reqs = calculate_research_requirements(empire)
     total_rp = empire.resourceProduction(fo.resourceType.research)
 
-    # get all usable researchable techs not already completed or queued for research
+    if total_rp <= 0: # No RP available - no research.
+        return
 
-    queued_techs = get_research_queue_techs()
-
-    def is_possible(tech_name):
-        return all([empire.getTechStatus(tech_name) == fo.techStatus.researchable,
-                   not tech_is_complete(tech_name),
-                   not exclude_tech(tech_name),
-                   tech_name not in queued_techs])
+    priorities = {}
+    for tech_name in fo.techs():
+        priority = get_priority(rng, tech_name)
+        if not tech_is_complete(tech_name) and priority >= 0:
+            turn_needed = float(research_reqs[tech_name][1]) / total_rp
+            priorities[tech_name] = float(priority) / turn_needed / max(1, (turn_needed - 12) / 4 + 1)**2
```

Comment:
As far as priorities for already researched techs go, I think it will simplify some things (at least for some changes I'm about to propose) if they're there, or at least can be.  It's easy enough to suppress printouts about them.

As for changing turn_needed, hang on a bit, because part of my proposed changes include tallying up actual min times from the prereqs, not just lumping all the cost together and doing a calc based on that.  So turns needed is not necessarily proportional to total cost.